lib/Driver/ToolChains.cpp

```diff
@@ -1270,8 +1293,37 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
     Arguments.push_back(context.Args.MakeArgString(context.OI.SDKPath));
   }
 
+  // Link the standard library.
   Arguments.push_back("-L");
-  Arguments.push_back(context.Args.MakeArgString(RuntimeLibPath));
+  if (context.Args.hasFlag(options::OPT_static_stdlib,
+                            options::OPT_no_static_stdlib,
+                            false)) {
+    SmallString<128> StaticRuntimeLibPath;
+    getRuntimeStaticLibraryPath(StaticRuntimeLibPath, context.Args, *this);
+    Arguments.push_back(context.Args.MakeArgString(StaticRuntimeLibPath));
+    // The following libraries are required to build a satisfactory
+    // static program
+    Arguments.push_back("-ldl");
+    Arguments.push_back("-lpthread");
+    Arguments.push_back("-lbsd");
+    Arguments.push_back("-licui18n");
+    Arguments.push_back("-licuuc");
+    // The runtime uses dlopen to look for the protocol conformances.
+    // Therefore, we need to ensure they appear in the dynamic table.
+    // This happens automatically for dynamically-linked programs, but
+    // in this case we have to take additional measures.
+    const char *dynamicList = context.getTemporaryFilePath("export-protocol-conformances", "");
+    const char dynamicListContents[] = "{\n.swift2_protocol_conformances_start;\n};";
```

Comment:
There's at least the `.swift2_type_metadata_start` table used for type-by-name lookup here: 

https://github.com/apple/swift/blob/master/stdlib/public/runtime/MetadataLookup.cpp#L51

and the reflection metadata @bitjammer 's been working on.

We set visibility in IRGen, but from the perspective that one module == one executable or dylib, so everything that's "public" in Swift ends up exported with protected visibility in the .o files. As we'd discussed previously, we could work around that at link time by telling `ld` to `--exclude-libs`, or be a bit smarter in IRGen and give public things only `hidden` visibility when we know we're building them for a .a target.